src/v/cloud_topics/throttler/throttler.cc

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_topics/throttler/throttler.h"
+
+#include "base/unreachable.h"
+#include "cloud_topics/core/event_filter.h"
+#include "cloud_topics/core/write_pipeline.h"
+#include "cloud_topics/core/write_request.h"
+#include "cloud_topics/logger.h"
+
+#include <seastar/core/loop.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/manual_clock.hh>
+#include <seastar/coroutine/as_future.hh>
+
+#include <exception>
+
+namespace experimental::cloud_topics {
+
+template<class Clock>
+throttler<Clock>::throttler(
```

Comment:
This is interesting but have 2 big questions:

1. Why throttler needs to copy requests internally and manage them instead of not progressing them to the next pipeline stage?
2. Can throttler introduce back-pressure? I believe that would be desirable but in this design it doesn't seem possible. Pipeline will keep accepting requests in _pending is if throttler is able to keep up (not guaranteed)? then it will drop requests when memory is full but if it doesn't keep up what happens?
3. In the current design with background threads there seems to be quite a bit of overhead. Can't we make items move through the pipeline eagerly? Rather than have "pipeline stages" "pull" elements make "pipeline" push elements to the first stage and the first stage will either drop/hold/push further the element. Out of the box we get backpressure support as well?. Reminds me of this talk https://dsdsd.da.cwi.nl/slides/dsdsd-duckdb-push-based-execution.pdf.